### ISSUE TITLE
Runaway Generation

### ISSUE THREAD

**Issue by sempervictus** [guidance-ai/llguidance#304]
I've almost completed the full guidance integration (have grammars, dynamic tool sleds, user-supplied constraints working) into `vllm.rs` but i'm kind of stuck in the expression language for natural output for expressions allowing a model to "just do its thing or call tools through enforced grammar of the tool call". It seems like anything i use in that field which allows generation of more than one token matching the pattern produces run-on effects.
```
start: TEXT
TEXT: /((?s).*?)/
```
or `TEXT: /((?s)[\x00-\u{10FFFF}])+(<[151645]>|<[151645]>|<[151643]>)?/` (extracting all EOS tokens from the tokenizer and composing them to try and bound the pattern) all result in the LLM _doing what is asked_ re the dynamic tool-call mechanics but then proceeding to vomit repetitious lines, individual tokens, often emoji's (q3 0.6 and q35 0.8 test models but stable when either constrained to tools or strict grammars or entirely unconstrained).

Have tried expressions such as:
```
start: text
text[stop=""]: /((?s).*?)/
```

Example of problematic output:

> ...
> Let me know if you'd like a more detailed explanation or examples! 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. 😊. ...

Something in the expression seems to be causing EOS to become "uncomputable" as _forced tool_ or _user-supplied_ constraints work great:

```
...   -H 'Content-Type: application/json' \
  -d '{
    "messages": [{"role":"user","content":"Generate a phone number"}],
    "constraint": "^number:\\s\\s\\d{3}-\\d{3}-\\d{4}\\ndo you want a sandwitch with that\\s\\S{6}",
    "constraint_type": "regex"
  }' | jq -r '.choices[0].message.content'
```
produces
```
number:  123-456-7890
do you want a sandwitch with that number?
```
_with_ the spelling mistake and formatted layout included (proving positional & content enforcement) and picking up constrained vs enforced generation of next-tokens after "that" correctly as well.

While the obvious thing would be just not apply a "free form constraint" the reason its needed is to create compositions such as:
```
start: (text | tool_call)+
text[stop=""]: /((?s).*?)/
tool_call: <[151657]> tool_obj <[151658]>
tool_obj: %json {"type":"object","properties":{"name":{"type":"string"},"arguments":{"type":"object"}},"required":["name","arguments"]}
json_array: "[" obj ("," obj)* "]"
obj_get_current_time: %json {"type":"object","properties":{},"required":[]}
obj: obj_fetch_url_via_curl | obj_fs_cat | obj_fs_ls | obj_get_current_time | obj_web_search_searxng
obj_fs_ls: %json {"type":"object","properties":{"path":{"type":"string","description":"The path of the directory to list"}},"required":["path"]}
obj_fetch_url_via_curl: %json {"type":"object","properties":{"url":{"type":"string","description":"The URL to scrape."},"proxy":{"type":"string","description":"The proxy URL in the format protocol://host:port."}},"required":["url"]}
obj_web_search_searxng: %json {"type":"object","properties":{"query":{"type":"string","description":"The query to search for."},"searxng":{"type":"string","description":"Optional searxng URL overriding the env var"}},"required":["query"]}
obj_fs_cat: %json {"type":"object","properties":{"path":{"type":"string","description":"The path of the file to read"}},"required":["path"]}
```
in order to inform precise tool-call+output turns which might iterate calls to different tools during the response in the sequence enforcing the constraint. Large models handle that sort of work very well, small ones and draft models less so. Guided control over _their_ generation at the throughput that they can produce seems a rather key capability toward production use of small models for doing actual work.

I get the sense i've gone too far into the weeds and am missing something obvious and apparent to sane people so figure i'd ask here - is there a forest i'm missing staring at tree bark or am i in fact not nuanced enough in how i'm expressing Lark constraints? The masking phase and inline per-token resampling (i actually haven't seen that happen in **some** time indicating i got the masking part right) work, at least to _constrain_ the input but is there something i might be missing in the mask generation/application logic that's causing this?

<details>

```rust

        let logits = if let Some(factory) = &self.llg_factory {
            let mut guidance_states = self.guidance_states.write();
            let mut guidance_failed = self.guidance_failed.write();
            let mut guidance_mismatch = self.guidance_mismatch.write();
            let mut modified = false;
            let vocab_size = logits.dim(1)?;

            // Only materialize logits on CPU if at least one constraint mask applies.
            // Collect masks first to minimize holding locks or complex logic inside the loop
            let mut masks: Vec<(usize, usize, SimpleVob)> = Vec::new(); // (seq_index, seq_id, mask)

            for (i, id) in seq_ids.iter().enumerate() {
                let sampling_params = match &seqs {
                    Seqs::SeqRefs(refs) => &refs[i].sampling_params,
                    Seqs::DecodeVec(vec) => &vec[i].sampling_params,
                };

                if guidance_failed.contains(id) {
                    continue;
                }

                // Use grammar directly from sampling_params
                let grammar = match sampling_params.grammar.as_ref() {
                    Some(g) => g,
                    None => continue,
                };

                let state = match guidance_states.entry(*id) {
                    Entry::Occupied(entry) => entry.into_mut(),
                    Entry::Vacant(entry) => {
                        match GuidanceState::new_from_grammar(factory.clone(), grammar) {
                            Ok(state) => entry.insert(state),
                            Err(err) => {
                                guidance_failed.insert(*id);
                                crate::log_warn!(
                                    "[Seq {}] Failed to create guidance state: {}. Disabling constraints for this sequence.",
                                    id,
                                    err
                                );
                                continue;
                            }
                        }
                    }
                };

                if let Ok(Some(mask)) = state.compute_mask() {
                    masks.push((i, *id, mask));
                    modified = true;
                }
            }

            if modified {
                // Convert to Vec, modify, and update logits
                let mut logits_vec = logits.flatten_all()?.to_vec1::<f32>()?;

                for (seq_idx, seq_id, mask) in masks {
                    let start = seq_idx * vocab_size;
                    let end = start + vocab_size;
                    let row = &mut logits_vec[start..end];
                    let mask_len = mask.len();

                    if mask_len == 0 {
                        if guidance_failed.insert(seq_id) {
                            crate::log_warn!(
                                "[Seq {}] Guidance mask length is 0. Disabling constraints for this sequence.",
                                seq_id
                            );
                        }
                        continue;
                    }

                    if mask_len != vocab_size && guidance_mismatch.insert(seq_id) {
                        crate::log_warn!(
                            "[Seq {}] Guidance mask size {} does not match vocab size {}. Clamping mask application.",
                            seq_id,
                            mask_len,
                            vocab_size
                        );
                    }

                    let apply_len = std::cmp::min(vocab_size, mask_len);
                    for tok in 0..apply_len {
                        if !mask.is_allowed(tok as u32) {
                            row[tok] = f32::NEG_INFINITY;
                        }
                    }
                    // Attempt to disallow generation past mask_len ... 
                    if mask_len < vocab_size {
                        for tok in mask_len..vocab_size {
                            row[tok] = f32::NEG_INFINITY;
                        }
                    }
                }
                Tensor::from_vec(logits_vec, logits.shape(), &self.device)?
            } else {
                logits
            }

        } else {
            logits
        };
```

</details>

None of the "we're in problem territory" log messages from the code above are emitted but the infinite stream of smiley-faces coming out of this do seem like they're laughing at me at this point 😉

**Comment by hudson-ai** [comment]
Hi @sempervictus it looks very much like an issue with the model "wanting" to produce an EOS but being unable to, causing it to spout nonsense. `llguidance` does not currently support multiple EOS tokens, so you have to either pass the "right" one at construction time, OR you can modify the grammar to explicitly include multiple. I believe you were on the right track with a fix when you wrote a regex that included the two EOS tokens, but you had a bit of a syntactic error.

In particular, the `<[N]>` syntax shouldn't be inside of the regex. It also needs to be put in a non-terminal (lowercase) rule.  Try
```lark
start: text_with_eos
text_with_eos: TEXT eos?
TEXT: /(?s:.*)/
eos: <[151645]> | <[151643]> // LHS has to be lower case -- literal tokens aren't allowed in TERMINAL rules
```

**Comment by sempervictus** [comment]
Thanks @hudson-ai - will take a shot at restoring the EOS extractors and constructing a proper EOS into a non-terminal rule.

While i've got your attention, any thoughts on the proper construction for `start: (text_with_eos | tool_call )+` logic? If we want the model to be able to to both, in chunks, would that make more sense as `start: (text_with_eos ? tool_call )+` and does the optimizer stick the `tool_call` in the front (of the processing order) anyway since its more constrained?

**Comment by hudson-ai** [comment]
> While i've got your attention, any thoughts on the proper construction for start: (text_with_eos | tool_call )+ logic? If we want the model to be able to to both, in chunks, would that make more sense as start: (text_with_eos ? tool_call )+ and does the optimizer stick the tool_call in the front (of the processing order) anyway since its more constrained?

I think that as long as the tool calls are indicated unambiguously with special tokens, this should work in a pretty straightforward manner (the grammar will be unambiguous). Otherwise, I think you'll have to be a bit careful and write a `lazy` lexeme that ensures that you unambiguously start a tool call as long as a particular string is generated. Can help with that if you try it out first and open an issue 😄

**Comment by hudson-ai** [comment]
And if that made no sense, let me know. I'll hopefully be more coherent tomorrow hahaha

**Comment by sempervictus** [comment]
Thanks 😁 - I'm fortunate enough to be surrounded by SME-madmen some of whom operate at your level so thankfully that actually made sense but the lower-case rule for EOS tokens works meaning i can almost put my PR to bed:

`$ aichat research Lark llguidance grammars and explain fast forward tokens along with usage in rust. Please make exactly two web search call and a current time call`

> `Call web_search_searxng {"query":"Lark grammar and fast forward tokens in Rust"}`
> `Processing 4096 tokens per chunk`
> `Call web_search_searxng {"query":"Lark grammar and fast forward tokens in Rust"}`
> `Processing 4096 tokens per chunk`
> `Call get_current_time {}`
> - **Lark Grammar and Fast Forward Tokens in Rust**: Lark grammars are structured to dynamically build token masks for every step of the decoding process, making them significantly slower compared to other implementations. This feature allows for efficient parsing of complex grammars.
> 
> - **Fast Forward Tokens**: These are a technique used in Lark grammars to append multiple tokens at once, enhancing the efficiency of parsing by reducing the number of tokenization steps.
> 
> - **Usage Examples**: In Rust, fast forward tokens can be implemented using Lark syntax, enabling developers to handle large and complex grammars with minimal computational overhead.
> 
> - **Current Time**: The current time is **Fri Mar  6 01:04:13 AM EST 2026**.
> 
> 

^^ for a 0.6B Q3 with 8K of context window allocated, not half bad.

<details>

> ...
> vllm-rs-svc2  | 2026-03-06T06:02:53.526799Z DEBUG vllm_rs::tools::schema: [llg] Using token IDs for start tag: <tool_call> -> <[151657]>
> vllm-rs-svc2  | 2026-03-06T06:02:53.526822Z DEBUG vllm_rs::tools::schema: [llg] Using token IDs for end tag: </tool_call> -> <[151658]>
> vllm-rs-svc2  | 2026-03-06T06:02:53.526832Z DEBUG vllm_rs::server::server: [llg] Built tool grammar from tools
> vllm-rs-svc2  | 2026-03-06T06:02:53.526834Z DEBUG vllm_rs::utils::guidance: [llg] compose_grammars() called: constraints=0
> vllm-rs-svc2  | 2026-03-06T06:02:53.526836Z DEBUG vllm_rs::utils::guidance: [llg] compose_grammars(): has_tools=true, tool_choice_required=false, forced_tool_name=None
> vllm-rs-svc2  | 2026-03-06T06:02:53.526858Z DEBUG vllm_rs::utils::guidance: [llg] compose_grammars() -> tool_call | text with EOS
> vllm-rs-svc2  | 2026-03-06T06:02:53.526862Z DEBUG vllm_rs::server::server: [llg] TopLevelGrammar for SamplingParams: TopLevelGrammar { grammars: [GrammarWithLexer [lark]], max_tokens: Some(16384) }
> vllm-rs-svc2  | 2026-03-06T06:02:53.526866Z DEBUG vllm_rs::server::server: [llg] Lark grammar string:
> vllm-rs-svc2  | start: ( text_with_eos | tool_call )+
> vllm-rs-svc2  | obj_fs_cat: %json {"type":"object","properties":{"path":{"type":"string","description":"The path of the file to read"}},"required":["path"]}
> vllm-rs-svc2  | obj_web_search_searxng: %json {"type":"object","properties":{"query":{"type":"string","description":"The query to search for."},"searxng":{"type":"string","description":"Optional searxng URL overriding the env var"}},"required":["query"]}
> vllm-rs-svc2  | ws: /[ \\t\\r\\n]+/
> vllm-rs-svc2  | obj_fetch_url_via_curl: %json {"type":"object","properties":{"url":{"type":"string","description":"The URL to scrape."},"proxy":{"type":"string","description":"The proxy URL in the format protocol://host:port."}},"required":["url"]}
> vllm-rs-svc2  | eos: <[151645]> | <[151645]> | <[151643]>
> vllm-rs-svc2  | json_array: "[" obj ("," obj)* "]"
> vllm-rs-svc2  | tool_call: <[151657]> tool_obj <[151658]>
> vllm-rs-svc2  | TEXT: /(?s:.*)/
> vllm-rs-svc2  | obj_fs_ls: %json {"type":"object","properties":{"path":{"type":"string","description":"The path of the directory to list"}},"required":["path"]}
> vllm-rs-svc2  | obj: obj_fetch_url_via_curl | obj_fs_cat | obj_fs_ls | obj_get_current_time | obj_web_search_searxng
> vllm-rs-svc2  | obj_get_current_time: %json {"type":"object","properties":{},"required":[]}
> vllm-rs-svc2  | tool_obj: %json {"type":"object","properties":{"name":{"type":"string"},"arguments":{"type":"object"}},"required":["name","arguments"]}
> vllm-rs-svc2  | text_with_eos: TEXT eos?
> vllm-rs-svc2  | 2026-03-06T06:02:53.526877Z  WARN vllm_rs::server::server: Tools enabled for request
> vllm-rs-svc2  | 2026-03-06T06:02:53.548913Z  WARN vllm_rs::core::engine: [Stream] New request [Seq_id 0, 1123 tokens] received! (session_id: None)
> vllm-rs-svc2  | 
> vllm-rs-svc2  | 2026-03-06T06:02:53.548958Z  INFO vllm_rs::server::parser: Tool parser selected: qwen (model_id=Qwen/Qwen3-0.6B, enforce_parser=none)
> vllm-rs-svc2  | 2026-03-06T06:02:53.582023Z  INFO vllm_rs::core::runner: User's thinking preference for reasoning models: None
> vllm-rs-svc2  | 2026-03-06T06:02:53.582044Z  WARN vllm_rs::core::runner: Using sampling from generation_config: temp=Some(0.6), top_k=Some(20), top_p=Some(0.95), freq_penalty=None, pres_penalty=None
> vllm-rs-svc2  | 2026-03-06T06:02:53.582053Z DEBUG vllm_rs::utils::guidance: [llg] GuidanceState::new_from_grammar() called
> vllm-rs-svc2  | 2026-03-06T06:02:53.582055Z DEBUG vllm_rs::utils::guidance: [llg] GuidanceState::new_from_grammar() called
> vllm-rs-svc2  | 2026-03-06T06:02:53.583336Z  INFO vllm_rs::utils::guidance: [llg] GuidanceState created successfully for grammar
> vllm-rs-svc2  | 2026-03-06T06:02:53.583350Z  INFO vllm_rs::utils::guidance: [llg] GuidanceState created successfully for grammar
> vllm-rs-svc2  | 2026-03-06T06:02:53.593246Z  INFO vllm_rs::server::parser: Tool call <tool_call> (151657) found, start buffering!
> vllm-rs-svc2  | 2026-03-06T06:02:53.700064Z  INFO vllm_rs::core::scheduler: [Seq 0] Detected </tool_call> token 151658, finishing for external handling
> vllm-rs-svc2  | 2026-03-06T06:02:53.700170Z  INFO vllm_rs::server::parser: Tool call buffering end, reached </tool_call> (151658)
> ...

sorry, raw logs - `vllm.rs` has IPC for front-end and model-runners hence the double log output for things which occur in the latter
> 

</details>

I think this syntax is _very_ well worth adding to the docs - lets just say a lot of brain-matter was spilled before filing this issue 😄 

If there are any LLG faculties you want to show off in a runtime, happy to implement or at least put on the TODO list while i'm elbow-deep in this stuff.
